journal/src/main/java/io/zeebe/journal/file/JournalSegment.java

```diff
@@ -139,12 +146,12 @@ public MappedJournalSegmentWriter writer() {
    */
   MappedJournalSegmentReader createReader() {
     checkOpen();
-    return new MappedJournalSegmentReader(file, this, index);
+    return new MappedJournalSegmentReader(buffer.asReadOnlyBuffer(), this, index);
   }
 
-  private MappedJournalSegmentWriter createWriter(
-      final JournalSegmentFile file, final int maxEntrySize) {
-    return new MappedJournalSegmentWriter(file, this, maxEntrySize, index);
+  private MappedJournalSegmentWriter createWriter(final int maxEntrySize) {
+    return new MappedJournalSegmentWriter(
+        (MappedByteBuffer) buffer.slice(), this, maxEntrySize, index);
```

Comment:
We have two options, one is what you implemented, the other is to pass readers as `buffer.asReadOnlyBuffer().position(0)`, and give the writers `buffer`, which avoids the cast and one extra allocation (but mainly avoids the cast :sweat_smile:).

imo I'd prefer this, as we don't recreate writers, but I won't insist on it.